### PR TITLE
Proposal: Introduce Array Nodes

### PR DESCRIPTION
**Pull request by katrogan**
Signed-off-by: Katrina Rogan <katroganGH@gmail.com>

# TL;DR
In the case of map tasks lets define the _parent task_ as the maptask definition in user code and the _subtask_ as what it yields. The task container target is insufficient to fully encapsulate the subtask because:

- both the parent task and subtask read from shared TaskTemplate fields such as the custom and task template version
- to support mapped pod tasks, the parent task needs to now construct spec from the K8sPod target alongside TaskTemplate fields

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1051
https://github.com/flyteorg/flyte/issues/1131

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by katrogan** [comment]
PTAL @wild-endeavor @EngHabu

File: protos/flyteidl/core/workflow.proto [hunk 1]
```diff
@@ -63,6 +63,26 @@ message WorkflowNode {
...
 
+// Configures an array job in which the node executes a task in parallel across a range of inputs.
+message ArrayNode {
+    // Reference to a registered task which should be called in parallel during node execution.
+    Identifier task_reference = 1;
+
+    // Defines the minimum number of instances to bring up concurrently at any given point. Note that this is an
+    // optimistic restriction and that, due to network partitioning or other failures, the actual number of currently
+    // running instances might be more. This has to be a positive number if assigned. Default value is size.
+    int64 parallelism = 2;
```
Comment (kumare3) [inline protos/flyteidl/core/workflow.proto:74]: I would call this concurrency

Comment (katrogan) [reply protos/flyteidl/core/workflow.proto]: done
